File: backend/app/utils/jwt.py
from datetime import datetime, timedelta
from typing import Optional
from jose import JWTError, jwt
from app.core.config import settings
from app.schemas.auth import TokenData


def create_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
    """Создает JWT access token"""
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode.update({"exp": expire, "type": "access"})
    encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return encoded_jwt


def create_refresh_token(data: dict) -> str:
    """Создает JWT refresh token"""
    to_encode = data.copy()
    expire = datetime.utcnow() + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    to_encode.update({"exp": expire, "type": "refresh"})
    encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return encoded_jwt


def verify_token(token: str, token_type: str = "access") -> Optional[TokenData]:
    """Верифицирует JWT token и возвращает данные"""
    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM])
        token_type_in_payload = payload.get("type")
        if token_type_in_payload != token_type:
            return None
        sub = payload.get("sub")
        email: str = payload.get("email")
        if sub is None or email is None:
            return None
        # Конвертируем sub (строка) обратно в int
        try:
            user_id: int = int(sub)
        except (ValueError, TypeError):
            return None
        return TokenData(user_id=user_id, email=email)
    except JWTError:
        return None

File: backend-node/src/utils/jwt.ts
import jwt from 'jsonwebtoken';
import { config } from '../core/config';

export interface TokenPayload {
  userId: number;
  email: string;
  type?: 'access' | 'refresh'; // Добавляем тип токена
}

export function createAccessToken(payload: TokenPayload): string {
  return jwt.sign(
    { ...payload, type: 'access' },
    config.secretKey,
    {
      algorithm: config.algorithm as jwt.Algorithm,
      expiresIn: `${config.accessTokenExpireMinutes}m`,
    }
  );
}

export function createRefreshToken(payload: TokenPayload): string {
  return jwt.sign(
    { ...payload, type: 'refresh' },
    config.secretKey,
    {
      algorithm: config.algorithm as jwt.Algorithm,
      expiresIn: `${config.refreshTokenExpireDays}d`,
    }
  );
}

export function verifyToken(token: string, expectedType?: 'access' | 'refresh'): TokenPayload {
  try {
    const decoded = jwt.verify(token, config.secretKey, {
      algorithms: [config.algorithm as jwt.Algorithm],
    }) as TokenPayload & { type?: string };

    // Проверяем тип токена, если указан
    if (expectedType && decoded.type !== expectedType) {
      throw new Error(`Invalid token type. Expected ${expectedType}, got ${decoded.type}`);
    }

    return decoded as TokenPayload;
  } catch (error) {
    if (error instanceof jwt.TokenExpiredError) {
      throw new Error('Token expired');
    }
    if (error instanceof jwt.JsonWebTokenError) {
      throw new Error('Invalid token');
    }
    throw error;
  }
}
